backend/filen/filen.go

```diff
@@ -0,0 +1,1084 @@
+	}
+	return f.moveWithParentUUID(ctx, item, newParentDir.GetUUID())
+}
+
+// moveWithParentUUID moves item to newParentUUID
+// using a simple filen.MoveItem operation
+func (f *Fs) moveWithParentUUID(ctx context.Context, item types.NonRootFileSystemObject, newParentUUID string) error {
+	err := f.filen.MoveItem(ctx, item, newParentUUID, true)
+	if err != nil {
+		return fmt.Errorf("failed to move file: %w : %w", err, fs.ErrorCantMove)
+	}
+
+	return nil
+}
+
+// rename moves item to newPath
+// using a simple Filen rename operation
+func (f *Fs) rename(ctx context.Context, item types.NonRootFileSystemObject, newPath string, newName string) error {
+	err := f.filen.Rename(ctx, item, newName)
+	if err != nil {
+		return fmt.Errorf("failed to rename item: %w : %w", err, fs.ErrorCantMove)
+	}
+	return nil
+}
+
+// moveFileObjIntoNewPath 'moves' an existing object into a new path
+// invalidating the previous object
+// and making a copy with the passed path
+//
+// this is to work around the fact that rclone expects to have to delete a file after moving
+func moveFileObjIntoNewPath(obj *File, newPath string) *File {
+	newFile := &File{
+		fs:   obj.fs,
+		path: newPath,
+		file: obj.file,
+	}
+	obj.file = nil
+	return newFile
+}
+
+// DirMove moves src, srcRemote to this remote at dstRemote
+// using server-side move operations.
+//
+// Will only be called if src.Fs().Name() == f.Name()
+//
+// If it isn't possible then return fs.ErrorCantDirMove
+//
+// If destination exists then return fs.ErrorDirExists
+func (f *Fs) DirMove(ctx context.Context, src fs.Fs, srcRemote, dstRemote string) error {
+
+	srcF, ok := src.(*Fs)
+	if !ok || srcF == nil {
+		return fs.ErrorCantDirMove
+	}
+	err := f.filen.Lock(ctx)
+	if err != nil {
+		return err
+	}
+	defer f.filen.Unlock()
+	g, gCtx := errgroup.WithContext(ctx)
+	var (
+		srcDirInt types.DirectoryInterface
+		dstDir    types.DirectoryInterface
+		srcPath   = srcF.resolvePath(srcF.Enc.FromStandardPath(srcRemote))
+		dstPath   = f.resolvePath(f.Enc.FromStandardPath(dstRemote))
+	)
+	if srcPath == dstPath {
+		return fs.ErrorDirExists
+	}
+
+	g.Go(func() error {
+		var err error
+		srcDirInt, err = srcF.filen.FindDirectory(gCtx, srcPath)
+		return err
+	})
+	g.Go(func() error {
+		var err error
+		dstDir, err = f.filen.FindDirectory(gCtx, dstPath)
+		return err
+	})
+
+	if err := g.Wait(); err != nil {
+		return err
+	}
+
+	if srcDirInt == nil {
+		return fs.ErrorDirNotFound
+	}
+
+	if dstDir != nil {
+		return f.dirMoveContents(ctx, srcDirInt, dstDir, srcPath, dstPath)
+	}
+
+	srcDir, ok := srcDirInt.(*types.Directory)
+	if !ok {
+		return fs.ErrorCantDirMove
+	}
+
+	return f.dirMoveEntireDir(ctx, srcDir, srcPath, dstPath)
+}
+
+// dirMoveContents moves the contents of srcDir to dstDir
+// used for the case where the target directory exists
+// recurses if needed
+func (f *Fs) dirMoveContents(ctx context.Context, srcDir, dstDir types.DirectoryInterface, srcPath, dstPath string) error {
```

Comment:
I think rclone should do this for you.

`DirMove` is allowed to fail if the destination exists.

That works fine for `rclone mount` and for `rclone move` it will engage backup code which does what this does.

https://github.com/rclone/rclone/blob/ff9cbab5fa36fef07f9237dcef073e16b0c72de8/fs/sync/sync.go#L1401

https://github.com/rclone/rclone/blob/ff9cbab5fa36fef07f9237dcef073e16b0c72de8/fs/sync/sync.go#L1421

So I think you should remove this function and just return `fs.ErrorDirExists`

NB the help for `DirMove` says

> If destination exists then return fs.ErrorDirExists

Reply:
Well that is annoying to have missed, I will simplify this.